packages/graphiql/src/state/GraphiQLSchemaProvider.tsx

```diff
@@ -0,0 +1,155 @@
+import React, { useCallback, useEffect } from 'react';
+import { GraphQLSchema } from 'graphql';
+import { defaultSchemaLoader } from './common';
+import { SchemaConfig } from './types';
+import {
+  SchemaAction,
+  SchemaActionTypes,
+  schemaRequestedAction,
+  schemaSucceededAction,
+  schemaErroredAction,
+} from './schemaActions';
+
+/**
+ * Initial State
+ */
+
+export type SchemaState = {
+  schema: GraphQLSchema | null;
+  isLoading: boolean;
+  errors: Error[] | null;
+  config: SchemaConfig;
+};
+
+export type SchemaReducer = React.Reducer<SchemaState, SchemaAction>;
+
+const isDev = window.location.hostname === 'localhost';
+
+const uri = isDev
+  ? 'http://localhost:8080/graphql'
+  : 'https://swapi-graphql.netlify.com/.netlify/functions/index';
+
+export const initialReducerState: SchemaState = {
+  isLoading: false,
+  config: {
+    uri,
+  },
+  schema: null,
+  errors: null,
+};
+
+export const getInitialState = (
+  initialState?: Partial<SchemaState>,
+): SchemaState => ({
+  ...initialReducerState,
+  ...initialState,
+});
+
+/**
+ * Context
+ */
+
+export type SchemaContextValue = SchemaState & ProjectHandlers;
+
+export const SchemaContext = React.createContext<SchemaContextValue>({
+  ...getInitialState(),
+  loadCurrentSchema: async () => undefined,
+  // loadSchema: async () => undefined,
+  dispatch: async () => undefined,
+  schemaLoader: defaultSchemaLoader,
+});
+
+export const useSchemaContext = () => React.useContext(SchemaContext);
+
+/**
+ * Action Types & Reducers
+ */
+
+export const schemaReducer: SchemaReducer = (state, action): SchemaState => {
+  switch (action.type) {
+    case SchemaActionTypes.SchemaChanged:
+      return {
+        ...state,
+        isLoading: true,
+        config: action.payload,
+      };
+    case SchemaActionTypes.SchemaRequested:
+      return {
+        ...state,
+        isLoading: true,
+      };
+    case SchemaActionTypes.SchemaSucceeded:
+      return {
+        ...state,
+        isLoading: false,
+        schema: action.payload,
+      };
+    case SchemaActionTypes.SchemaErrored:
+      return {
+        ...state,
+        isLoading: false,
+        errors: state.errors
+          ? [...state.errors, action.payload]
+          : [action.payload],
+      };
+    default: {
+      return state;
+    }
+  }
+};
+
+/**
+ * Provider
+ */
+
+export type SchemaProviderProps = {
+  config?: SchemaConfig;
+  schemaLoader?: typeof defaultSchemaLoader;
+  children?: any;
+};
+
+export type ProjectHandlers = {
+  // loadSchema: (state: SchemaState, config: SchemaConfig) => Promise<void>;
+  loadCurrentSchema: (state: SchemaState) => Promise<void>;
+  schemaLoader: typeof defaultSchemaLoader;
+  dispatch: React.Dispatch<SchemaAction>;
+};
+
+export function SchemaProvider({
```

Comment:
Would it not be more within react conventions to expose this context as `SchemaContext.Provider`?

Reply:
i think you're supposed to expose the context as in the result of createContext. as for context providers, usually the pattern is to return SchemaContext.Provider, which we're doing here